code/modules/mob/living/silicon/robot/laws.dm

```diff
@@ -43,45 +27,29 @@
 		to_chat(who, "<b>Remember, you are not bound to any AI, you are not required to listen to them.</b>")
 
 
-/mob/living/silicon/robot/proc/lawsync()
+/mob/living/silicon/robot/lawsync()
+	if(!connected_ai?.laws)
+		return
+
 	laws_sanity_check()
-	var/datum/ai_laws/master = connected_ai ? connected_ai.laws : null
-	var/temp
-	if (master)
-		laws.devillaws.len = master.devillaws.len
-		for (var/index = 1, index <= master.devillaws.len, index++)
-			temp = master.devillaws[index]
-			if (length(temp) > 0)
-				laws.devillaws[index] = temp
-
-		laws.ion.len = master.ion.len
-		for (var/index = 1, index <= master.ion.len, index++)
-			temp = master.ion[index]
-			if (length(temp) > 0)
-				laws.ion[index] = temp
-
-		laws.hacked.len = master.hacked.len
-		for (var/index = 1, index <= master.hacked.len, index++)
-			temp = master.hacked[index]
-			if (length(temp) > 0)
-				laws.hacked[index] = temp
-
-		if(master.zeroth_borg) //If the AI has a defined law zero specifically for its borgs, give it that one, otherwise give it the same one. --NEO
-			temp = master.zeroth_borg
-		else
-			temp = master.zeroth
-		laws.zeroth = temp
-
-		laws.inherent.len = master.inherent.len
-		for (var/index = 1, index <= master.inherent.len, index++)
-			temp = master.inherent[index]
-			if (length(temp) > 0)
-				laws.inherent[index] = temp
-
-		laws.supplied.len = master.supplied.len
-		for (var/index = 1, index <= master.supplied.len, index++)
-			temp = master.supplied[index]
-			if (length(temp) > 0)
-				laws.supplied[index] = temp
+
+	if(deployed)
+		laws = connected_ai.laws
+
+	else
+		if(!lawupdate)
+			return
+
+		laws.zeroth = connected_ai.laws.zeroth_borg || connected_ai.laws.zeroth
+
+		laws.hacked = connected_ai.laws.hacked
+
+		laws.ion = connected_ai.laws.ion
+
+		laws.inherent = connected_ai.laws.inherent
+
+		laws.supplied = connected_ai.laws.supplied
+
+	post_lawchange(FALSE)
```

Comment:
This proc is being called by `lawsync()` in one level, and calling it in another. Seems a bit dangerous, no?

Reply:
`ai/post_lawchange()` -> `ai/lawsync()` -> `robot/lawsync()` -> `silicon/post_lawchange()`